### PR TITLE
Allow passing in optional bounds to the pipeline

### PR DESCRIPTION
**Pull request by JasonGross**

### CODE REVIEW

**Comment by andres-erbsen** [comment]
~~~~~~~~~~~~~~
File "./src/Experiments/SimplyTypedArithmetic.v", line 7532, characters 11-25:
Error:
In environment
N, R, N', machine_wordsize : Z
n := Z.to_nat (Qceiling (inject_Z (Z.log2_up N) / inject_Z machine_wordsize))
 : nat
bound := r[0 ~> 2 ^ machine_wordsize - 1]%zrange : zrange
rw := rweight (inject_Z machine_wordsize)
 : Expr (type.type_primitive type.nat -> type.type_primitive type.Z)
rw_half := rweight (inject_Z machine_wordsize / inject_Z 2)
 : Expr (type.type_primitive type.nat -> type.type_primitive type.Z)
relax_zrange := relax_zrange_of_machine_wordsize : zrange -> option zrange
rv : ?T
rop :
Expr
  (type.type_primitive type.Z * type.type_primitive type.Z ->
   type.type_primitive type.Z)
E : ?T0
Hrop : ?T1
HE : ?T2
The term "(bound, bound)" has type "(zrange * zrange)%type"
while it is expected to have type
 "partial.data (type.type_primitive type.Z * type.type_primitive type.Z)".
Command exited with non-zero status 1
src/Experiments/SimplyTypedArithmetic (real: 180.29, user: 179.43, sys: 0.39, mem: 1152496 ko)
make: *** [src/Experiments/SimplyTypedArithmetic.vo] Error 1
~~~~~~~~~~~

**Comment by JasonGross** [comment]
Should compile now (it does locally, at least); I've `git commit --amend`ed with the change
```diff
diff --git a/src/Experiments/SimplyTypedArithmetic.v b/src/Experiments/SimplyTypedArithmetic.v
index e51f44a..3378f70 100644
--- a/src/Experiments/SimplyTypedArithmetic.v
+++ b/src/Experiments/SimplyTypedArithmetic.v
@@ -7495,7 +7495,7 @@ Module MontgomeryReduction.
             (machine_wordsize : Z).
 
     Let n : nat := Z.to_nat (Qceiling ((Z.log2_up N) / machine_wordsize)).
-    Let bound := r[0 ~> (2^machine_wordsize - 1)%Z]%zrange.
+    Let bound := Some r[0 ~> (2^machine_wordsize - 1)%Z]%zrange.
 
     Definition relax_zrange_of_machine_wordsize
       := relax_zrange_gen [1; machine_wordsize / 2; machine_wordsize; 2 * machine_wordsize; 4 * machine_wordsize]%Z.
```

**Comment by JasonGross** [comment]
> nothing looks obviously off

The most "off" thing is that I changed the way boundedness is defined in `Module Ring` to preserve things lining up judgmentally (but this is part what I think you'll be rewriting soon, and, also, I think that parameterizing over the notion of boundedness should also work equally well).
The second biggest change is that I corrected the definition of `is_bounded_by` so that `None` is considered bottom rather than top, in the case of `list`.
Basically everything else is boiler-plate / piping through changes.
